src/main/java/org/opensearch/knn/index/mapper/EngineFieldMapper.java

```
@@ -223,6 +223,23 @@ private EngineFieldMapper(
         this.vectorValidator = knnLibraryIndexingContext.getVectorValidator();
     }
 
+    private VectorSimilarityFunction findBestMatchingVectorSimilarityFunction(final SpaceType spaceType) {
+        if (indexCreatedVersion.onOrAfter(Version.V_3_0_0)) {
+            // We need to find the best matching similarity function and not just save DEFAULT space type after 3.0.
+            // This is required for memory optimized search where utilizing .vec file to retrieve vectors.
+            // During the retrieval, it will locate similarity function from the meta info. Without this best effort, always the default
+            // similarity function will be used even when other space type is configured in a mapping.
+            // However, for keeping the backward compatibility, we only apply this to indices created after 3.0+.
+            try {
+                return spaceType.getKnnVectorSimilarityFunction().getVectorSimilarityFunction();
+            } catch (Exception e) {
```

Why not catch the unsupported operation exception?

---

Hamming space type is the only space that is throwing `IllegalStateException` at the moment.
I can replace Exception to `IllegalStateException`, but `getVectorSimilarityFunction()` can actually throw arbitrary exceptions in there. So I've decided to have it Exception instead of enumerating all possible exceptions to be thrown.

---

Lets not ignore this exception. Please throw this exception out.

---

Just to provide some context why Default space type was used:
@jmazanec15 and @0ctopus13prime BTW this is not only about Hamming distance. For older indices with Nmslib engine a user may have set L1/linf as space type. Hence, default space type was chosen when we moved to Lucene based vector field for our native engines.

---

I think we should suppress the exception and fallback to the default one.
The purpose of this check is to test whether it could find the best matching similarity function in Lucene.
And Hamming is just throwing IllegalArgumentException when try to get vector similarity function. 
If we just rethrow everything we got, then user cannot create an index with hamming distance.
I understand that snatching all exception then suppress might be uncomfortable, I think this is the way to achieve the goal while minimizing maintenance efforts. One day, a FAISS unique space type can be added where it throws UnsupportedOperationException.